docs: putting test into each chapter of the tutorial (#1647)
canonical/operator: docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/create-a-minimal-kubernetes-charm.md

**Congratulations, you've successfully created a minimal Kubernetes charm!** 

(write-unit-tests-for-your-charm)=
## Write unit tests for your charm

> When you're writing a charm, you will want to ensure that it will behave reliably as intended.

Reviewer: ```suggestion
When you're writing a charm, you will want to ensure that it will behave as intended.
```

> For example, that the various components -- relation data, Pebble services, or configuration files -- all behave as expected in response to an event.

Reviewer: ```suggestion
For example, you'll want to check that the various components -- relation data, Pebble services, or configuration files -- all behave as expected in response to an event.
```

> You can ensure all this by writing a rich battery of unit tests. In the context of a charm we recommended using [`pytest`](https://pytest.org/) (but [`unittest`](https://docs.python.org/3/library/unittest.html) can also be used) and especially Ops's built-in testing library --  [](ops_testing). We will be using the Python testing tool [`tox`](https://tox.wiki/en/4.14.2/index.html) to automate our testing and set up our testing environment.

Reviewer: ```suggestion
You can ensure all this by writing a rich battery of unit tests. In the context of a charm, we recommended using [`pytest`](https://pytest.org/) ([`unittest`](https://docs.python.org/3/library/unittest.html) can also be used) with [](ops_testing), the framework for state-transition testing in Ops.

We'll also use the Python testing tool [`tox`](https://tox.wiki/en/4.14.2/index.html) to automate our testing and set up our testing environment.
```

> In this chapter you will write a test to check that Pebble is configured as expected.

Reviewer: ```suggestion
In this section we'll write a test to check that Pebble is configured as expected.
```

### Prepare your test environment

> Create a file called `tox.ini` in your charm project's root directory and add the following to configure your test environment:

Reviewer: ```suggestion
Create a file called `tox.ini` in your project root directory and add the following configuration:
```

```
[tox]
no_package = True
skip_missing_interpreters = True
min_version = 4.0.0
env_list = unit

[vars]
src_path = {tox_root}/src
tests_path = {tox_root}/tests

[testenv]
set_env =
    PYTHONPATH = {tox_root}/lib:{[vars]src_path}
    PYTHONBREAKPOINT=pdb.set_trace
    PY_COLORS=1
pass_env =
    PYTHONPATH
    CHARM_BUILD_DIR
    MODEL_SETTINGS

[testenv:unit]
description = Run unit tests
deps =
    pytest
    cosl
    coverage[toml]
    ops[testing]
    -r {tox_root}/requirements.txt
commands =
    coverage run --source={[vars]src_path} \
                 -m pytest \
                 --tb native \
                 -v \
                 -s \
                 {posargs} \
                 {[vars]tests_path}/unit
    coverage report
```
> Read more: [`tox.ini`](https://tox.wiki/en/latest/config.html#tox-ini)

> ```{tip}
> 
> Once you've mastered the basics, you can speed things up by navigating to your empty charm project directory and running `charmcraft init --profile kubernetes`. This will create the `tox.ini` config file for you.
> 
> ```

Reviewer: ```suggestion
If you used `charmcraft init --profile kubernetes` at the beginning of your project, you will already have the `tox.ini` file.
```

### Prepare your test directory

> In your project root, create a `tests/unit` directory:

Reviewer: ```suggestion
In your project root directory, create directory for the unit test:
```

```text
mkdir -p tests/unit
```

> ### Write your test

Reviewer: ```suggestion
### Write a test
```

> In your `tests/unit` directory, create a new file `test_charm.py` and add the test below. This test will check the behaviour of the `_on_demo_server_pebble_ready` function that you set up in a previous chapter. It will first set up the test context, then define the input state, then run the action and, finally, check if the results match the expected values.

Reviewer: ```suggestion
In your `tests/unit` directory, create a new file called `test_charm.py` and add the test below. This test will check the behaviour of the `_on_demo_server_pebble_ready` function that you set up earlier. The test will first set up a context, then define the input state, run the action, and check whether the results match the expected values.
```

from charm import FastAPIDemoCharm

def test_pebble_layer():
    ctx = testing.Context(FastAPIDemoCharm)
    container = testing.Container(name="demo-server", can_connect=True)
    state_in = testing.State(
        containers={container},
        leader=True,
    )
    state_out = ctx.run(ctx.on.pebble_ready(container), state_in)
    # Expected plan after Pebble ready with default config
    expected_plan = {
        "services": {
            "fastapi-service": {
                "override": "replace",
                "summary": "fastapi demo",
                "command": "uvicorn api_demo_server.app:app --host=0.0.0.0 --port=8000",
                "startup": "enabled",
                # Since the environment is empty, Layer.to_dict() will not
                # include it.
            }
        }
    }

    # Check that we have the plan we expected:
    assert state_out.get_container(container.name).plan == expected_plan
    # Check the unit status is active
    assert state_out.unit_status == testing.ActiveStatus()
    # Check the service was started:
    assert container.service_statuses["fastapi-service"] == ops.pebble.ServiceStatus.ACTIVE
```

### Run the test

> In your Multipass Ubuntu VM shell, run your test as below:

Reviewer: ```suggestion
In your Multipass Ubuntu VM shell, run your test:
```

```text
ubuntu@charm-dev:~/fastapi-demo$ tox -e unit     
```

> You should get an output similar to the one below:

Reviewer: ```suggestion
The result should be similar to the following output:
```

=================================================================== test session starts ===================================================================
platform linux -- Python 3.13.2, pytest-8.3.5, pluggy-1.5.0 -- /home/ubuntu/juju-sdk-tutorial-k8s/.tox/unit/bin/python
cachedir: .tox/unit/.pytest_cache
rootdir: /home/ubuntu/juju-sdk-tutorial-k8s
configfile: pyproject.toml
collected 1 item

tests/unit/test_charm.py::test_pebble_layer PASSED

==================================================================== 1 passed in 0.11s ====================================================================
unit: commands[1]> coverage report
Name           Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------------
src/charm.py      18      0      0      0   100%
----------------------------------------------------------
TOTAL             18      0      0      0   100%
  unit: OK (2.12=setup[1.82]+cmd[0.26,0.04] seconds)
  congratulations :) (2.14 seconds)
```

Congratulations, you have written your first unit test!

(write-integration-tests-for-your-charm)=
## Write integration tests for your charm

> A charm should function correctly not just in a mocked environment but also in a real deployment.

Reviewer: ```suggestion
A charm should function correctly not just in a mocked environment, but also in a real deployment.
```

For example, it should be able to pack, deploy, and integrate without throwing exceptions or getting stuck in a `waiting` or a `blocked` status -- that is, it should correctly reach a status of `active` or `idle`.

You can ensure this by writing integration tests for your charm. In the charming world, these are usually written with the [`pytest-operator`](https://github.com/charmed-kubernetes/pytest-operator) library.

> In this chapter you will write a small integration test to check that the charm packs and deploys correctly.

Reviewer: ```suggestion
In this section we'll write a small integration test to check that the charm packs and deploys correctly.
```

### Prepare your test environment

In your `tox.ini` file, add the following new environment:

```
[testenv:integration]
description = Run integration tests
deps =
    pytest
    juju
    pytest-operator
    -r {tox_root}/requirements.txt
commands =
    pytest -v \
           -s \
           --tb native \
           --log-cli-level=INFO \
           {posargs} \
           {[vars]tests_path}/integration
```

> If you used `charmcraft init --profile kubernetes` in previous steps to bootstrap your project, the `testenv:integration` section is already in the `tox.ini` config file.

Reviewer: ```suggestion
If you used `charmcraft init --profile kubernetes` at the beginning of your project, the `testenv:integration` section is already in the `tox.ini` file.
```

### Prepare your test directory

> In your project root, create a `tests/integration` directory:

Reviewer: ```suggestion
In your project root directory, create a directory for the integration test:
```

```text
mkdir -p tests/integration
```

### Write and run a pack-and-deploy integration test

> Let's begin with the simplest possible integration test, a [smoke test](https://en.wikipedia.org/wiki/Smoke_testing_(software)). This test will build and deploy the charm and verify that the installation hooks finish without any error. 

Reviewer: ```suggestion
Let's begin with the simplest possible integration test, a [smoke test](https://en.wikipedia.org/wiki/Smoke_testing_(software)). This test will build and deploy the charm, then verify that the installation event is handled without errors.
```

> In your `tests/integration` directory, create a file `test_charm.py` and add the following test case:

Reviewer: ```suggestion
In your `tests/integration` directory, create a file called `test_charm.py` and add the following test:
```

from pytest_operator.plugin import OpsTest

logger = logging.getLogger(__name__)

METADATA = yaml.safe_load(Path("./charmcraft.yaml").read_text())
APP_NAME = METADATA["name"]

@pytest.mark.abort_on_fail
async def test_build_and_deploy(ops_test: OpsTest):
    """Build the charm-under-test and deploy it together with related charms.

    Assert on the unit status before any relations/configurations take place.
    """
    # Build and deploy charm from local source folder
    charm = await ops_test.build_charm(".")
    resources = {
        "demo-server-image": METADATA["resources"]["demo-server-image"]["upstream-source"]
    }

    await asyncio.gather(
        ops_test.model.deploy(charm, resources=resources, application_name=APP_NAME),
        ops_test.model.wait_for_idle(
            apps=[APP_NAME], status="active", raise_on_blocked=False, timeout=300
        ),
    )
```

In your Multipass Ubuntu VM, run the test:

```bash
tox -e integration
```

The test takes some time to run as the `pytest-operator` running in the background will add a new model to an existing cluster (whose presence it assumes). If successful, it'll verify that your charm can pack and deploy as expected.

> You should get an output similar to the one below:

Reviewer: ```suggestion
The result should be similar to the following output:
```

**Congratulations, you've successfully created a minimal Kubernetes charm!** 

(write-unit-tests-for-your-charm)=
## Write unit tests for your charm

When you're writing a charm, you will want to ensure that it will behave reliably as intended.

For example, that the various components -- relation data, Pebble services, or configuration files -- all behave as expected in response to an event.

> You can ensure all this by writing a rich battery of unit tests. In the context of a charm we recommended using [`pytest`](https://pytest.org/) (but [`unittest`](https://docs.python.org/3/library/unittest.html) can also be used) and especially the operator framework's built-in testing library --  [](ops_testing_harness). We will be using the Python testing tool [`tox`](https://tox.wiki/en/4.14.2/index.html) to automate our testing and set up our testing environment.

Reviewer 2: ```suggestion
You can ensure all this by writing a rich battery of unit tests. In the context of a charm we recommended using [`pytest`](https://pytest.org/) (but [`unittest`](https://docs.python.org/3/library/unittest.html) can also be used) and especially Ops's built-in testing library --  [](ops_testing). We will be using the Python testing tool [`tox`](https://tox.wiki/en/4.14.2/index.html) to automate our testing and set up our testing environment.
```

I think we're moving away from "operator framework" to just "Ops". Also, we should link to Scenario, not Harness.

<!-- TODO

This chapter and the next should be removed and the content spread throughout
each of the previous chapters. Each time a feature is added, a unit and
integration test should also be added. At the end of each chapter, before
manually checking the feature works, the user should run the tests to make sure
that they pass.

-->

> 

Reviewer 2: We can remove this now. It ended up sitting in my PR for ages and then not very long in main, oh well.

> In this chapter you will write a scenario test to check that the `get_db_info` action that you defined in an earlier chapter behaves as expected.

Reviewer 2: This is the first chapter, right? So this should be more something like:

```suggestion
In this chapter you will write a test to check that Pebble is configured as expected.
```

You can ensure this by writing integration tests for your charm. In the charming world, these are usually written with the [`pytest-operator`](https://github.com/charmed-kubernetes/pytest-operator) library.

In this chapter you will write a small integration test to check that the charm packs and deploys correctly.

### Prepare your test environment

In your `tox.ini` file, add the following new environment:

```
[testenv:integration]
description = Run integration tests
deps =
    pytest
    juju
    pytest-operator
    -r {tox_root}/requirements.txt
commands =
    pytest -v \
           -s \
           --tb native \
           --log-cli-level=INFO \
           {posargs} \
           {[vars]tests_path}/integration
```

If you used `charmcraft init --profile kubernetes` in previous steps to bootstrap your project, the `testenv:integration` section is already in the `tox.ini` config file.

### Prepare your test directory

Create  a  `tests/integration` directory:

```text
mkdir ~/fastapi-demo/tests/integration

> ```

Reviewer 2: I think for consistency we should do what you did above with "In your project root...". This is better in some ways, since you can run it anywhere, but I think for a beginner being told where to start is actually easier.

Author: For consistency, I'll use the "project root" approach - When following other tutorials, I sometimes find that I don't want or can't use what the tutorial proposed as a working directory, so even if it's `mkdir -p ~/proejct/root/some/path/`, I can't always copy-paste. For example maybe the users don't want to use their home directory as a working directory.
